src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,410 @@
+            //  - elements from left and right, in case of a tuple literal
+            //  - left and right, in case of a tuple or nullable tuple type
+            //
+            // But if neither is nullable, then just produce the inner sequence.
+            //
+            // Note: all the temps are created in a single bucket (rather than different scopes of applicability) for simplicity
+
+            var outerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+            var innerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+
+            BoundExpression leftHasValue;
+            BoundExpression leftValue;
+
+            var coreLeft = WithoutImplicitNullableConversions(left);
+            var isLeftNullable = coreLeft.Type.IsNullableType();
+            if (isLeftNullable)
+            {
+                BoundExpression savedLeft = EvaluateSideEffectingArgumentToTemp(VisitExpression(left), outerEffects, ref initialEffectsAndTemps.temps);
+                leftHasValue = MakeHasValueTemp(savedLeft, initialEffectsAndTemps.temps, outerEffects);
+                leftValue = MakeValueOrDefaultTemp(savedLeft, initialEffectsAndTemps.temps, innerEffects);
+                inLeftLiteral = false;
+            }
+            else
+            {
+                leftHasValue = MakeBooleanConstant(left.Syntax, true);
+                leftValue = SaveTupleIfNotLiteral(coreLeft, ref inLeftLiteral, initialEffectsAndTemps, isRight: false);
+            }
+
+            BoundExpression rightHasValue;
+            BoundExpression rightValue;
+
+            var coreRight = WithoutImplicitNullableConversions(right);
+            var isRightNullable = coreRight.Type.IsNullableType();
+            if (isRightNullable)
+            {
+                BoundExpression savedRight = EvaluateSideEffectingArgumentToTemp(VisitExpression(right), outerEffects, ref initialEffectsAndTemps.temps);
+                rightHasValue = MakeNullableHasValue(savedRight.Syntax, savedRight); // no need for local for right.HasValue since used once
+                rightValue = MakeValueOrDefaultTemp(savedRight, initialEffectsAndTemps.temps, innerEffects);
+                inRightLiteral = false;
+            }
+            else
+            {
+                rightHasValue = MakeBooleanConstant(right.Syntax, true);
+                rightValue = SaveTupleIfNotLiteral(coreRight, ref inRightLiteral, initialEffectsAndTemps, isRight: true);
+            }
+
+            BoundExpression logicalExpression = RewriteNonNullableNestedTupleOperators(operators, leftValue, rightValue, boolType, initialEffectsAndTemps.temps, innerEffects, initialEffectsAndTemps, operatorKind, inLeftLiteral, inRightLiteral);
+
+            BoundExpression innerSequence = _factory.Sequence(ImmutableArray<LocalSymbol>.Empty, innerEffects.ToImmutableAndFree(), logicalExpression);
+
+            if (!isLeftNullable && !isRightNullable)
+            {
+                return innerSequence;
```

Comment:
Are we dropping side-effects from `initialEffectsAndTemps`? Consider asserting if that should be empty.

Reply:
The side-effects from `initialEffectsAndTemps` are always applied at the top-level (see `VisitTupleBinaryOperator` on line 36).
This case is just an optimization of `outer sequence a few lines below, when we know that both `leftHasValue` and `rightHasValue` are true.
I'll add a comment.


---
In reply to: [170397917](https://github.com/dotnet/roslyn/pull/23356#discussion_r170397917) [](ancestors = 170397917)